consumer/src/main/java/com/flipkart/varadhi/consumer/delivery/MessageDelivery.java

```diff
@@ -0,0 +1,71 @@
+package com.flipkart.varadhi.consumer.delivery;
+
+import com.flipkart.varadhi.entities.Endpoint;
+import com.flipkart.varadhi.entities.Message;
+import com.flipkart.varadhi.exceptions.NotImplementedException;
+import com.google.common.collect.Multimap;
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+
+public interface MessageDelivery {
+    static MessageDelivery of(Endpoint endpoint) {
+        return switch (endpoint.getProtocol()) {
+            case HTTP1_1 -> new HttpMessageDelivery(endpoint);
+            case HTTP2 -> throw new NotImplementedException("HTTP2 is not supported yet");
+            default -> throw new IllegalArgumentException("Unsupported protocol: " + endpoint.getProtocol());
+        };
+    }
+
+    CompletableFuture<DeliveryResponse> deliver(Message message)
+            throws Exception;
+
+    class HttpMessageDelivery implements MessageDelivery {
+        private final Endpoint.HttpEndpoint endpoint;
+        private final HttpClient httpClient;
+
+        public HttpMessageDelivery(Endpoint endpoint) {
+            this.endpoint = (Endpoint.HttpEndpoint) endpoint;
+            this.httpClient = HttpClient.newBuilder()
+                    .version(this.endpoint.isHttp2Supported() ? HttpClient.Version.HTTP_2 : HttpClient.Version.HTTP_1_1)
+                    .connectTimeout(Duration.ofMillis(this.endpoint.getConnectTimeoutMs()))
+                    .followRedirects(HttpClient.Redirect.NORMAL)
+                    .build();
+        }
+
+        @Override
+        public CompletableFuture<DeliveryResponse> deliver(Message message) throws Exception {
+            HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
+                    .uri(endpoint.getUrl().toURI())
```

Comment:
The fact that we are forced to do this `endpoint.getUrl().toURI())` , raises a question. Should we also just keep the URI object in the Endpoint object instead of URL?

Reply:
sure, I can make this change